src/persist-client/src/write.rs

```diff
@@ -132,35 +132,63 @@ where
     {
         trace!("WriteHandle::append lower={:?} upper={:?}", lower, upper);
 
+        let batch = match self
+            .write_batch(updates, lower.clone(), upper.clone())
+            .await
+        {
+            Ok(batch) => batch,
+            Err(invalid_usage) => return Ok(Err(invalid_usage)),
+        };
+
+        self.append_core(batch, lower, upper).await
+    }
+
+    /// Appends the given batch of updates to this shard and downgrades this
+    /// handle's upper to `upper`.
+    ///
+    /// The innermost `Result` is `Ok` if the updates were successfully written.
+    /// If not, an `Upper` err containing the current writer upper is returned.
+    /// If that happens, we also update our local `upper` to match the current
+    /// upper. This is useful in cases where a timeout happens in between a
+    /// successful write and returning that to the client.
+    ///
+    /// In contrast to [Self::compare_and_append_core], multiple [WriteHandle]s
+    /// may be used concurrently to write to the same shard, but in this case,
+    /// the data being written must be identical (in the sense of
+    /// "definite"-ness). It's intended for replicated use by source ingestion,
+    /// sinks, etc.
+    ///
+    /// A `upper` of the empty antichain "finishes" this shard, promising that
+    /// no more data is ever incoming.
+    ///
+    /// The batch may be empty, which allows for downgrading `upper` to
+    /// communicate progress. It is possible to heartbeat a writer lease by
+    /// calling this with `upper` equal to `self.upper()` and an empty `updates`
+    /// (making the call a no-op).
+    ///
+    /// The clunky multi-level Result is to enable more obvious error handling
+    /// in the caller. See <http://sled.rs/errors.html> for details.
+    pub async fn append_core(
```

Comment:
I _think_ we should be able to write this method entirely in terms of compare_and_append_core (i.e. append is something that I'm pretty sure could live completely outside of the persist crate because it's implementable in terms of the _public_ compare_and_append). this would have been clunky before this PR (it would have had to write out the blob again on every expected_upper mismatch), but now IIUC it should be totally doable

Reply:
> I think we should be able to write this method entirely in terms of compare_and_append_core (i.e. append is something that I'm pretty sure could live completely outside of the persist crate because it's implementable in terms of the public compare_and_append). this would have been clunky before this PR (it would have had to write out the blob again on every expected_upper mismatch), but now IIUC it should be totally doable

done